tree/tree/src/InternalTreeUtils.cxx

```diff
@@ -409,35 +410,48 @@ std::vector<std::unique_ptr<TChain>> MakeFriends(const ROOT::TreeUtils::RFriendI
+/// \brief Recursively expand the glob to take care of potential wildcard
+/// specials for subdirectories in the glob.
+/// \param[in] l The list of full paths to files.
 /// \param[in] glob The glob to expand.
-/// \throws std::runtime_error If the directory part of the glob refers to a
+/// \throws std::runtime_error If the directory parts of the glob refer to a
 ///         path that cannot be opened.
-/// \return A vector of strings, the fully expanded paths to the files referred
-///         to by the glob.
 ///
-/// The two parts of the glob (directory, fileglob) are separated. The directory
-/// is first expanded via TSystem::ExpandPathName then opened via
-/// TSystem::OpenDirectory. If the directory can be opened, then the remaining
-/// fileglob is used as regex expression (via TRegexp) and the function stores
-/// those files in the directory that match the regex.
-std::vector<std::string> ExpandGlob(const std::string &glob)
+/// If the glob contains a wildcard special for subdirectories, the three parts
+/// of the glob (directory, subdirectoryglob, remainder) are separated.
+/// Otherwise the glob is expanded to (directory, fileglob).
+/// The directory is first expanded via TSystem::ExpandPathName then opened via
+/// TSystem::OpenDirectory. If the directory can be opened, then current
+/// glob is used as regex expression (via TRegexp) to find subdirectories or
+/// store those files in the directory that match the regex.
+void RecursiveGlob(TList &l, const std::string &glob)
 {
-   // TODO: Also implement this pattern in some other free function
-   // This is practically Python's os.path.split
    std::string dirname;
-   std::string basename;
+   std::string basename; // current glob to expand, could be a directory or file.
+   std::string remainder;
 
-   const auto slashpos = glob.rfind('/');
+   const char *wildcardSpecials = "[]*?";
 
-   if (slashpos != std::string::npos) {
-      // Separate the glob into its two components
-      dirname = glob.substr(0, slashpos);
-      basename = glob.substr(slashpos + 1);
+   const auto wildcardPos = glob.find_first_of(wildcardSpecials);
+   // Get the closest slash, to the left of the first wildcard
+   const auto slashLPos = glob.rfind("/", wildcardPos);
+   // Get the closest slash, to the right of the first wildcard
+   const auto slashRPos = glob.find("/", wildcardPos);
+
+   if (slashLPos != std::string::npos) {
+      // Separate the base directory im the glob.
+      dirname = glob.substr(0, slashLPos);
    } else {
       // There is no directory component in the glob, use the CWD
       dirname = gSystem->UnixPathName(gSystem->WorkingDirectory());
-      basename = glob;
+   }
+
+   // Seperate the subdirectory and/or file component.
+   if (slashRPos != std::string::npos) {
+      basename = glob.substr(slashLPos + 1, slashRPos - slashLPos - 1);
+      remainder = glob.substr(slashRPos + 1);
+   } else {
+      basename = glob.substr(slashLPos + 1);
```

Comment:
Isn't a possible that we get here with `slashLPos == std::string::npos` ?

Reply:
Good point, turns out it was still extracting the correct basename because `std::string::npos` is currently defined as -1. So with `slashLPos + 1`, the basename is set to the full glob or until the first `/` if `slashRPos != std::string::npos`. To make this more explicit and to be sure  that `slashLPos + 1 == 0` if `slashLPos == std::string::npos`, I've added the line `slashLPos = -1`.

Reply:
Thanks.